Style inheritance

zarov opened the pull request:
Two commits, that try to add `Style` inheritance. Still need to fill in some proper documentation, but we can discuss it in the meantime.

#### feat(style): add Style inheritance

It is now possible to specify a `Style` parent for another `Style`. That
way, it is possible to define a global `Style` for a `ColorLayer`, while
defining one with property specific to a `FeatureGeometry` for example.
The `Style` of the `FeatureGeometry` will then only have the minimum
requirement, and it will read missing properties from its parent. It is
a dynamic reading, so it the parent values are changing, it is changing
as well for the child style.

This is a proposition for improvement in #1318.

BREAKING CHANGE: `Style#text.halo` is not available anymore, and each
property of it as been place in the `text` object, with the `halo`
prefix: `haloWidth`, `haloColor` and `haloBlur`.

#### chore: update tests and examples to support the new Style inheritance

gchoqueux commented on examples/source_file_geojson_raster.html:
```diff
@@ -68,6 +68,15 @@
             // Convert by iTowns
             itowns.Fetcher.json('https://raw.githubusercontent.com/gregoiredavid/france-geojson/master/departements/09-ariege/departement-09-ariege.geojson')
                 .then(function _(geojson) {
+                    optionsGeoJsonParser.style = new itowns.Style({
```
If we add a style in layer instance, the inheritance will work?
I would prefer add the global style in Layer because is more intuitive

zarov replied:
I agree, and this should be the case most of the times. 

However, here we have a special case: the data is parsed before being passed to its `FileSource`, using `GeoJsonParser`. In it, the `Style` assignment is done, but at this time it still don't know to which layer it will be attached, so it doesn't have a `parentStyle` passed in it. Normally, it is done in `DataSourceProvider`, when we don't parse the content ourselves. But here, we have to specify the `style` option.

gchoqueux replied:
This is only for `FileSource` ?
It's possible to set `parent` style after parse data?

zarov replied:
This is only for `FileSource` yes. Not currently, but I'll try to think of a solution for it.

zarov replied:
I added a proposition in 63ec609: it modifies `DataSourceProvider`, but to me it seems correct to do this.

gchoqueux commented on examples/source_file_geojson_raster.html:
```diff
@@ -77,48 +86,44 @@
+                .then(texture => {
+                    pyoStyle = new itowns.Style({
+                        fill: {
+                            pattern: texture.image,
```
We could think to fetch image in Style

zarov replied:
Done in 3d10626

gchoqueux commented on src/Provider/DataSourceProvider.js:
```diff
@@ -9,6 +9,33 @@ const error = (err, source) => {
 
+function convertSourceData(data, extDest, layer) {
+    // Add missing parent style
+    if (data.isFeatureCollection) {
```
Could you move this in new `FeatureCollection` function?

gchoqueux commented on src/Provider/DataSourceProvider.js:
```diff
@@ -9,6 +9,33 @@ const error = (err, source) => {
+            }
+
+            if (parentStyle) {
+                f.geometries.forEach((g) => {
```
Could you move this in new `Feature` function?

gchoqueux commented on src/Provider/DataSourceProvider.js:
```diff
@@ -9,6 +9,33 @@ const error = (err, source) => {
+            if (f.style && f.style.isStyle) {
+                parentStyle = f.style;
+
+                if (layer.style && !f.style.parent.isStyle) {
```
you check if `style.parent` is defined to avoid reassigning?

zarov replied:
Yes, it was to set it only once if missing.

gchoqueux reviewed (commented):
May be add `inheritance` property to Style instead of `parent`.
because it avoids to traverse all child styles.

```js
const parentStyle = new Style(c);
const inheritance = { parent: parentStyle };
const style_a = new Style(a, inheritance);
const style_b = new Style(b, inheritance);

inheritance.parent = new Style(v);
```

> Add missing parent style

is only for `FileSource` case?
if so then you could do in `layer.whenReady` property.

zarov commented:
> May be add `inheritance` property to Style instead of `parent`.
> because it avoids to traverse all child styles.
> 
> ```js
> const parentStyle = new Style(c);
> const inheritance = { parent: parentStyle };
> const style_a = new Style(a, inheritance);
> const style_b = new Style(b, inheritance);
> 
> inheritance.parent = new Style(v);
> ```

Why not, but I don't see where this could be used ? The `inheritance` object isn't kept, no ?

> > Add missing parent style
> 
> is only for `FileSource` case?
> if so then you could do in `layer.whenReady` property.

In this case yes, but I don't see how I could use `layer.whenReady` ?

gchoqueux commented:
> Why not, but I don't see where this could be used ? The inheritance object isn't kept, no ?

May be like this:
```js
style_b.inheritance.parent = new Style(v);
```

> In this case yes, but I don't see how I could use layer.whenReady ?

May be here:

https://github.com/iTowns/itowns/blob/master/src/Layer/Layer.js#L100-L103

zarov commented:
I updated from your comments, but I can't find a correct way to use the `inheritance` property, sorry.

gchoqueux commented on src/Core/Feature.js:
```diff
@@ -132,6 +132,12 @@ export class FeatureGeometry {
     }
+
+    setParentStyle(parentStyle) {
+        if (this.properties.style && this.properties.style.isStyle && !this.properties.style.parent.isStyle) {
```
is necessary to check `isStyle`?

zarov replied:
It is not, I removed it, before we needed to do that but indeed it is not the case anymore.

gchoqueux commented on src/Core/Feature.js:
```diff
@@ -223,6 +229,15 @@ class Feature {
+
+    setParentStyle(parentStyle) {
+        if (this.style && this.style.isStyle) {
+            this.style.parent = parentStyle;
```
may be add inheritance when you instance  Geometry?

zarov replied:
I didn't, because `Feature.style` is always present, and we don't always want a style on a geometry (see vector tiles). However, I set the inheritance right after instancing it in `GeoJsonParser`.

gchoqueux commented on src/Layer/Layer.js:
```diff
@@ -99,7 +99,13 @@ class Layer extends THREE.EventDispatcher {
-                return parseSourceData(this.source.fetchedData, this.source.extent, this);
+                return parseSourceData(this.source.fetchedData, this.source.extent, this).then((parsedData) => {
+                    // Set the missing style if it is a FileSource and a FeatureCollection
+                    if (parsedData.isFeatureCollection) {
```
is necessary?

zarov replied:
No, removed it !

gchoqueux commented on src/Parser/GeoJsonParser.js:
```diff
@@ -42,54 +42,54 @@ const toFeature = {
         feature.updateExtent(geometry);
     },
-    multi(type, feature, crsIn, coordsIn, filteringExtent, setAltitude, properties) {
+    multi(type, feature, crsIn, coordsIn, filteringExtent, setAltitude, properties, parentStyle) {
```
could you pass parent style when you instance feature?

zarov replied:
Done

zarov committed:
feat(style): add Style inheritance

It is now possible to specify a `Style` parent for another `Style`. That
way, it is possible to define a global `Style` for a `ColorLayer`, while
defining one with property specific to a `FeatureGeometry` for example.
The `Style` of the `FeatureGeometry` will then only have the minimum
requirement, and it will read missing properties from its parent. It is
a dynamic reading, so it the parent values are changing, it is changing
as well for the child style.

This is a proposition for improvement in #1318.

BREAKING CHANGE: `Style#text.halo` is not available anymore, and each
property of it as been place in the `text` object, with the `halo`
prefix: `haloWidth`, `haloColor` and `haloBlur`.

zarov committed:
chore: update tests and examples to support the new Style inheritance

zarov committed:
feature(style): allow URL of a fill.pattern to be used

If a fill.pattern is an URL, it is used to fetch the pattern, using
Fetcher.texture().
